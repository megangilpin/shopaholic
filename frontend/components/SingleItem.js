import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Error from './ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head';
import formatMoney from '../lib/formatMoney';
import AddToCart from './AddToCart';

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    // width: auto;
    // height: 85vh;
    // margin: 0 auto;
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 4rem 4rem;
    line-height: 1;
    color: ${props => props.theme.darkGrey};
  }
  .item_title {
    text-transform: uppercase;
    letter-spacing: .4rem;
    font-weight: 400;
    font-size: 4rem;
    margin-bottom: 1.5rem;
  }
  .description_title {
    text-transform: uppercase;
    letter-spacing: .4rem;
    font-weight: 600;
    font-size: 2rem;
    color: ${props => props.theme.darkGrey};
    margin-top: 4rem;
    margin-bottom: 1.5rem;
  }
  .description {
    font-weight: 400;
    font-size: 1.5rem;
    color: ${props => props.theme.darkPurple};
    margin: 0;
  }
  button {
    height: 7rem;
    display: flex;
    align-items: center;
    margin-top: 4rem;
    padding: 1.7rem 2rem;
    text-transform: uppercase;
    letter-spacing: .2rem;
    font-weight: 600;
    font-size: 1.5rem;
    background: none;
    border: 0;
    border-radius: 10px;
    cursor: pointer;
    color:  ${props => props.theme.grey};
    transition: all .3s;
    &:hover {
      box-shadow: -6px -6px 10px white,
      1px 1px 5px rgba(0, 0, 0, 0.1);
      -webkit-transform: scale(1.1);
      -ms-transform: scale(1.1);
      transform: scale(1.1);
    }
    &:active 
    {
      border: none;
      box-shadow: inset -6px -6px 10px white, inset
      1px 1px 5px rgba(0, 0, 0, 0.1);
    }
  }
  hr {
    border: 0;
    height: 0;
    border-top: 1px solid ${props => props.theme.lightPurple};
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      price
      description
      largeImage
    }
  }
`;
class SingleItem extends Component {
  render() {
    return (
      <Query
        query={SINGLE_ITEM_QUERY}
        variables={{
          id: this.props.id,
        }}
      >
        {({ error, loading, data }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          if (!data.item) return <p>No Item Found for {this.props.id}</p>;
          const item = data.item;
          return (
            <SingleItemStyles>
              <Head>
                <title>Sick Fits | {item.title}</title>
              </Head>
              <img src={item.largeImage} alt={item.title} />
              <div className="details">
                <h2 className="item_title">{item.title}</h2>
                <hr></hr>
                <h2>{formatMoney(item.price)}</h2>
                <p className="description_title">Details:</p>
                <p className="description">{item.description}</p>
                <AddToCart id={item.id} />
              </div>
            </SingleItemStyles>
          );
        }}
      </Query>
    );
  }
}

export default SingleItem;
export { SINGLE_ITEM_QUERY };
