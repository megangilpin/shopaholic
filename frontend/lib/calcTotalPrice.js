export default function calcTotalPrice(cart) {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    let amount = cartItem.item.price
    return tally + cartItem.quantity * amount;
  }, 0);
}
