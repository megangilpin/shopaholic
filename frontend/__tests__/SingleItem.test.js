import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeItem } from "../lib/testUtils";


describe('<SingleItem/>', () => {
  // needs a mock provider because we are wrapping our entire application in apollo provider
  it('renders with proper data', async () => {
    const mocks = [
      {
        // when someone makes a request with this query and variable combo
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        // return this fake data (mocked data)
        result: {
          data: {
            item: fakeItem(),
          },
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    // console.log(wrapper.debug())
    expect(wrapper.text()).toContain('Loading...');
    // mocks waiting for data abd puts the call at the end of the stack
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
  });
  
  it('Errors with a not found item', async () =>{
    const mocks = [{
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: {
          errors: [{ message: 'Items Not Found!'}],
        },
      },
    ]
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const item = wrapper.find('[data-test="graphql-error"]');
    expect(item.text()).toContain('Items Not Found!');
    expect(toJSON(item)).toMatchSnapshot();
  })
})