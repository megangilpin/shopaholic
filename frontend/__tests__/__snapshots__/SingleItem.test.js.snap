// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> Errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items Not Found!
</p>
`;

exports[`<SingleItem/> renders with proper data 1`] = `
Array [
  <h2
    className="item_title"
  >
    dogs are best
  </h2>,
  <h2>
    $50
  </h2>,
]
`;

exports[`<SingleItem/> renders with proper data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<SingleItem/> renders with proper data 3`] = `
Array [
  <p
    className="description_title"
  >
    Details:
  </p>,
  <p
    className="description"
  >
    dogs
  </p>,
]
`;
